@ojmel 

The unit tests are failing because of the TypeScript issue:

```
Error: routes/saveWSIAnnotation.ts(111,42): error TS2339: Property 'id' does not exist on type '{}'.
Error: routes/saveWSIAnnotation.ts(113,59): error TS2339: Property 'id' does not exist on type '{}'.
Error: Process completed with exit code 2.
```


https://github.com/stjude/proteinpaint/actions/runs/24693905587/job/72222150007?pr=4437

--- a/client/plots/wsiviewer/interactions/WSIViewerInteractions.ts
+++ b/client/plots/wsiviewer/interactions/WSIViewerInteractions.ts
@@ -122,13 +122,20 @@ export class WSIViewerInteractions {
 				const idx = currentIndex
 				if (event.key == '.') {
 					//Do not react if at the last annotation
-					if (currentIndex == tileSelections.length) return
-					currentIndex += 1
+					if (currentIndex == tileSelections.length) {
+						currentIndex = 0
+					} else {
+						currentIndex += 1
+					}
 				}
 				if (event.key == ',') {
 					//Do not react if at the starting annotation
-					if (currentIndex === 0) return
-					currentIndex -= 1
+					if (currentIndex === 0) {
+						// If at the starting tileselection, find the the most recent Annotation by checking for timestamp property
+						currentIndex = tileSelections.findIndex(ts => 'timestamp' in ts) || 0
+					} else {
+						currentIndex -= 1
+					}
 				}
 
 				if (idx !== currentIndex) {

--- a/client/plots/wsiviewer/interactions/WSIViewerInteractions.ts
+++ b/client/plots/wsiviewer/interactions/WSIViewerInteractions.ts
@@ -196,6 +196,8 @@ export class WSIViewerInteractions {
 					return
 				}
 				if (shortcuts.includes(event.code) && !settings.isSavingAnnotation) {
+					// TODO: Issue where if you press shortcuts too fast, annotation table doesnt get updated with correct class
+					//My guess is that if you press this fast enough, saveAnnotation doesnt update fast enough
 					wsiApp.app.dispatch({
 						type: 'plot_edit',
 						id: wsiApp.id,
@@ -218,16 +220,7 @@ export class WSIViewerInteractions {
 
 					// Persist and finalize via helper
 					await this.saveAndFinalizeAnnotation(wsiApp, sessionWSImage, currentIndex, selectedClassId, aiProjectID)
-					wsiApp.app.dispatch({
-						type: 'plot_edit',
-						id: wsiApp.id,
-						config: {
-							settings: {
-								isSavingAnnotation: false,
-								changeTrigger: Date.now()
-							}
-						}
-					})
+
 					return
 				}
 			})
@@ -631,6 +624,7 @@ export class WSIViewerInteractions {
 					renderWSIViewer: false,
 					changeTrigger: Date.now(),
 					activeAnnotation: 0,
+					isSavingAnnotation: false,
 					sessionsTileSelection: sessionsTileSelection
 				}
 			}

--- a/client/plots/wsiviewer/viewModel/SessionWSImage.ts
+++ b/client/plots/wsiviewer/viewModel/SessionWSImage.ts
@@ -80,12 +80,13 @@ export class SessionWSImage extends WSImage {
 				{ value: '' }
 			]
 		})
-
+		console.log('new')
 		const annotationsRows: any[] = annotations.map((annotation, i) => {
 			const idx = sessionsRows.length + predictionRows.length + i // Continue index
 			const color = sessionWSImage.classes?.find(c => c.label === annotation.class)?.color
 			const firstCell: any = { value: idx }
 			firstCell.origBackground = idx === selectedTileIndex ? selectedColor : ''
+			console.log(annotation.class, color)
 			return [
 				firstCell,
 				{ value: annotation.zoomCoordinates },

--- a/server/routes/saveWSIAnnotation.ts
+++ b/server/routes/saveWSIAnnotation.ts
@@ -91,7 +91,29 @@ function validateQuery(ds: any, connection: Database.Database) {
 			}
 
 			const imageId = imageRow.id
-
+			//Checking for duplicate annotation based on projectId, imageId and coordinates
+			const duplicateCheckSql = `
+				SELECT id
+				FROM project_annotations
+				WHERE project_id = ?
+				  AND image_id = ?
+				  AND coordinates = ?
+				LIMIT 1
+			`
+			const duplicateCheckStmt = connection.prepare(duplicateCheckSql)
+			const duplicateRow = duplicateCheckStmt.get(projectId, imageId, coords)
+			if (duplicateRow) {
+				const deleteDuplicateSql = `
+					DELETE FROM project_annotations
+					WHERE id = ?
+				`
+				const deleteDuplicateStmt = connection.prepare(deleteDuplicateSql)
+				deleteDuplicateStmt.run(duplicateRow.id)
+				console.log(
+					`Deleted duplicate annotation with id=${duplicateRow.id} for project_id=${projectId}, image_id=${imageId}.`
+				)
+				//Gotta retrieve the info for what was deleted
+			}
 			const insertSql = `
 				INSERT INTO project_annotations (
 					project_id, user_id, coordinates, timestamp, status, class_id, image_id

--- a/server/routes/saveWSIAnnotation.ts
+++ b/server/routes/saveWSIAnnotation.ts
@@ -101,7 +101,7 @@ function validateQuery(ds: any, connection: Database.Database) {
 				LIMIT 1
 			`
 			const duplicateCheckStmt = connection.prepare(duplicateCheckSql)
-			const duplicateRow = duplicateCheckStmt.get(projectId, imageId, coords)
+			const duplicateRow = duplicateCheckStmt.get(projectId, imageId, coords) as { id: number } | undefined
 			if (duplicateRow) {
 				const deleteDuplicateSql = `
 					DELETE FROM project_annotations

